presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/StaticHiveCluster.java

```diff
@@ -65,13 +67,29 @@ public StaticHiveCluster(List<URI> metastoreUris, String metastoreUsername, Hive
     public HiveMetastoreClient createMetastoreClient()
             throws TException
     {
-        List<HostAndPort> metastores = new ArrayList<>(addresses);
-        Collections.shuffle(metastores.subList(1, metastores.size()));
-
         TException lastException = null;
-        for (HostAndPort metastore : metastores) {
+
+        for (HostAndPort metastore : addresses) {
             try {
-                HiveMetastoreClient client = clientFactory.create(metastore);
+                HiveMetastoreClient client = new FailureAwareHiveMetaStoreClient(clientFactory.create(metastore), new ResponseHandle()
+                {
+                    @Override
+                    public void success()
+                    {
+                        // do nothing
+                    }
+
+                    @Override
+                    public void failed(TException e)
+                    {
+                        synchronized (addresses) {
+                            if (addresses.size() > 1) {
+                                addresses.remove(metastore);
+                                addresses.add(metastore);
```

Comment:
This is somehow "poor man" approach. Notice that in case of having two metastore client where one always fails and two concurrent queries. We could switch metastore twice and hit the wrong metastore again... and possible again.

Maybe we could sort them by number of failures and select the one which has lowest failures ratio. Also we would need to reset that failures count every time metastore returns successfully.

We could have additional abstraction for that like `MetastoreAddressSupplier`

Reply:
Good idea. 

> Maybe we could sort them by number of failures and select the one which has lowest failures ratio. Also we would need to reset that failures count every time metastore returns successfully

If we have 2 metastores then non-working metastore would be hit every second request in this approach.

Maybe we should have two lists, one with working metastores, the other with failed metastores. For a new request we select a random metastore from the "working" list. If "working" list is empty, then we select one from "failed" list. A metastore from "failed" list would be put in "working" list after some time (this time could increase exponentially when consecutive requests fail).

What do you think?

Reply:
> If we have 2 metastores then non-working metastore would be hit every second request in this approach.
Could you elaborate this?

How about each metastore has a backoff duration that will increase exponentially when request fail? this should work the same as the two list approach.

Reply:
> How about each metastore has a backoff duration that will increase exponentially when request fail? this should work the same as the two list approach.

I think this is similar to what I've described. You probably need to blacklist such failed metastores for a duration of backoff.

Reply:
I implemented a version using Backoff.